course5/homework10/linked_list.h

```diff
@@ -0,0 +1,221 @@
+#pragma once
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+
+namespace linked_list
+{
+
+// This is a very helpful website: https://www.cs.northwestern.edu/~riesbeck/programming/c++/stl-iterator-define.html#TOC23
+template<class T> class LinkedListIter;
+
+template<class T, class Allocator = std::allocator<T>>
+class LinkedList
+{
+public:
+    friend class LinkedListIter<T>;
+    typedef LinkedListIter<T> iterator;
+
+    // Basic assignment constructors
+    LinkedList() = default;
+    LinkedList(const T& val) // TODO T&& val ?
```

Comment:
`push_front` on an empty list doesn't seem to cause memory leaks, but once I use this constructor, I start leaking memory (I think this is it, at least). I don't think I should need to define a destructor (all the linked list contains is some `unique_ptr` -- shouldn't it just work?).